Files.Launcher/Win32API.cs

```diff
@@ -21,553 +21,553 @@
+						}
+						break;
+					}
+				}
+			});
+		}
+
+		public static string GenerateUniquePath(string path)
+		{
+			string uniquePath = path;
+
+			if (File.Exists(path))
+			{
+				string nameWithoutExt = Path.GetFileNameWithoutExtension(path);
+				string extension = Path.GetExtension(path);
+				string directory = Path.GetDirectoryName(path);
+
+				for (ushort count = 1; File.Exists(uniquePath); count++)
+				{
+					if (Regex.IsMatch(nameWithoutExt, @".*\(\d+\)$"))
+					{
+						uniquePath = Path.Combine(directory, $"{nameWithoutExt.Substring(0, nameWithoutExt.LastIndexOf("(", StringComparison.InvariantCultureIgnoreCase))}({count}){extension}");
+					}
+					else
+					{
+						uniquePath = Path.Combine(directory, $"{nameWithoutExt} ({count}){extension}");
+					}
+				}
+			}
+			else if (Directory.Exists(path))
+			{
+				string directory = Path.GetDirectoryName(path);
+				string Name = Path.GetFileName(path);
+
+				for (ushort Count = 1; Directory.Exists(uniquePath); Count++)
+				{
+					if (Regex.IsMatch(Name, @".*\(\d+\)$"))
+					{
+						uniquePath = Path.Combine(directory, $"{Name.Substring(0, Name.LastIndexOf("(", StringComparison.InvariantCultureIgnoreCase))}({Count})");
+					}
+					else
+					{
+						uniquePath = Path.Combine(directory, $"{Name} ({Count})");
+					}
+				}
+			}
+
+			return uniquePath;
+		}
+
+		/// <summary>
+		/// Gets file path from file FRN
+		/// </summary>
+		/// <param name="frn">File reference number</param>
+		/// <param name="volumeHint">Drive containing the file (e.g. "C:\")</param>
+		/// <returns>File path or null</returns>
+		public static string PathFromFileId(ulong frn, string volumeHint)
+		{
+			string volumePath = Path.GetPathRoot(volumeHint);
+			using var volumeHandle = Kernel32.CreateFile(volumePath, Kernel32.FileAccess.GENERIC_READ, FileShare.Read, null, FileMode.Open, FileFlagsAndAttributes.FILE_FLAG_BACKUP_SEMANTICS);
+			if (volumeHandle.IsInvalid) return null;
+			var fileId = new Kernel32.FILE_ID_DESCRIPTOR() { Type = 0, Id = new Kernel32.FILE_ID_DESCRIPTOR.DUMMYUNIONNAME() { FileId = (long)frn } };
+			fileId.dwSize = (uint)Marshal.SizeOf(fileId);
+			using var hFile = Kernel32.OpenFileById(volumeHandle, fileId, Kernel32.FileAccess.GENERIC_READ, FileShare.Read, null, FileFlagsAndAttributes.FILE_FLAG_BACKUP_SEMANTICS);
+			if (hFile.IsInvalid) return null;
+			var sb = new StringBuilder(4096);
+			var ret = Kernel32.GetFinalPathNameByHandle(hFile, sb, 4095, 0);
+			return (ret != 0) ? sb.ToString() : null;
+		}
+
+		public class Win32Window : IWin32Window
+		{
+			public IntPtr Handle { get; set; }
+
+			public static Win32Window FromLong(long hwnd)
+			{
+				return new Win32Window() { Handle = new IntPtr(hwnd) };
+			}
+		}
+
+		// Get information from recycle bin.
+		[DllImport(Lib.Shell32, SetLastError = false, CharSet = CharSet.Auto)]
+		public static extern int SHQueryRecycleBin(string pszRootPath,
+			ref SHQUERYRBINFO pSHQueryRBInfo);
+	}
 }
```

Comment:
Can you revert the changes to this file?

Reply:
I couldn't rollback for some reason, but I took the file from the main branch and overwrote it, then committed the change. Does that make a difference on your side?